test-snapshots/snapshots-output/Snapshot.ConstructorAccessor.ss

```diff
@@ -0,0 +1,97 @@
+#!r6rs
+#!chezscheme
+(library
+  (Snapshot.ConstructorAccessor lib)
+  (export
+    First*
+    First-value0
+    First?
+    Foo
+    Foo*
+    Foo-value0
+    Foo-value1
+    Foo-value2
+    Foo?
+    HasArgs
+    HasArgs*
+    HasArgs-value0
+    HasArgs-value1
+    HasArgs-value2
+    HasArgs?
+    Last*
+    Last-value0
+    Last?
+    NoArgs
+    NoArgs?
+    test1
+    test2
+    test3
+    test4
+    test5
+    x
+    y
+    z)
+  (import
+    (prefix (chezscheme) scm:)
+    (prefix (_Chez_Runtime lib) rt:))
+
+  (scm:define-record-type (First$ First* First?)
+    (scm:fields (scm:immutable value0 First-value0)))
+
+  (scm:define-record-type (Last$ Last* Last?)
+    (scm:fields (scm:immutable value0 Last-value0)))
+
+  (scm:define NoArgs
+    (scm:quote NoArgs))
+
+  (scm:define NoArgs?
+    (scm:lambda (v)
+      (scm:eq? (scm:quote NoArgs) v)))
+
+  (scm:define-record-type (HasArgs$ HasArgs* HasArgs?)
+    (scm:fields (scm:immutable value0 HasArgs-value0) (scm:immutable value1 HasArgs-value1) (scm:immutable value2 HasArgs-value2)))
+
+  (scm:define HasArgs
+    (scm:lambda (value0)
+      (scm:lambda (value1)
+        (scm:lambda (value2)
+          (HasArgs* value0 value1 value2)))))
+
+  (scm:define-record-type (Foo$ Foo* Foo?)
+    (scm:fields (scm:immutable value0 Foo-value0) (scm:immutable value1 Foo-value1) (scm:immutable value2 Foo-value2)))
+
+  (scm:define Foo
+    (scm:lambda (value0)
+      (scm:lambda (value1)
+        (scm:lambda (value2)
+          (Foo* value0 value1 value2)))))
```

Comment:
Seems like the order of the outputted definitions don't match the source order. I would expect the order to be:
- `Foo` (e.g. `Foo$`, `Foo`)
- `NoArgs` (e.g. `NoArgs`, `NoArgs?`)
- `HasArgs`
- `First`
- `Last`

Reply:
The ordering is provided by BackendModule.bindings. Tracing the idents as follows

```haskell
bindingsIdents :: Array String                                                                                                                                           
bindingsIdents = map (coerce <<< Tuple.fst) $ Array.concatMap (_.bindings) bindings
```

Produces

```
[37 of 42] Building Snapshot.ConstructorAccessor                                                                                                                             
["First","Last","NoArgs","HasArgs","Foo","z","y","x","test5","test4","test3","test2","test1"]
```

I'm not sure how we can get the source code order.

Reply:
I've opened https://github.com/aristanetworks/purescript-backend-optimizer/issues/46 upstream. Maybe it can be fixed there? Maybe not?